Allow metrics to be configurable by level

revans2 opened the pull request:
This fixes #1626 

The goal of this is to reduce the number of metrics that we report by default and also to allow the user to disable most metrics if they hit a situation like the one in the issue being fixed.

I realize that this is a huge change, in terms of the files and lines that it touches.

revans2 commented:
build

jlowe commented on shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala:
```diff
@@ -96,12 +93,15 @@ case class GpuBroadcastHashJoinExec(
-    "streamTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "stream time"),
-    "joinTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "join time"),
-    "filterTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "filter time"))
+  override lazy val outputRowsLevel: MetricsLevel = ESSENTIAL_LEVEL
+  override lazy val outputBatchesLevel: MetricsLevel = MODERATE_LEVEL
```
Curious, why are these `lazy`?  Comment applies to many other places where this was copied.

revans2 replied:
I made it lazy so on the chance that we might want to calculate it later. If you want me to switch it over to just be a val I can.

jlowe replied:
Yeah these should all not be lazy.  If some exec in the future wants to calculate this dynamically then it can do a lazy override, but no need to apply this to all cases.  I'm also not sure why we would want a metric to sometimes be essential vs. moderate etc.

revans2 replied:
You cannot do a lazy override to a non-lazy value.  They have to match, but we can cross that bridge when we come to it.

jlowe commented on shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/GpuBroadcastHashJoinExec.scala:
```diff
@@ -5,7 +5,7 @@
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *   http://www.apache.org/licenses/LICENSE-2.0
```
Intentional change?

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExec.scala:
```diff
@@ -47,15 +67,75 @@ object GpuMetricNames {
+  val DESCRIPTION_BUILD_TIME = "build time"
+  val DESCRIPTION_STREAM_TIME = "stream time"
+
+  // TODO these end up being ESSENTIAL_LEVEL do we really want that?
```
Part of me thinks no GPU metric should be essential, only the Spark built-ins are, but I can see the argument that some GPU-specific metrics on some execs could be considered essential.  Curious what others think.

revans2 replied:
I traced through a number of places for the metrics to see if they matched the Spark versions, and these I didn't find the time to do yet.  I'll look at them and see if I need to refactor it.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvtxWithMetrics.scala:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, NVIDIA CORPORATION.
+ * Copyright (c) 2019,2021, NVIDIA CORPORATION.
```
This file was changed in 2020, copyright should have been updated.
```suggestion
 * Copyright (c) 2019-2021 NVIDIA CORPORATION.
```

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -410,6 +429,21 @@ object RapidsConf {
+  val METRICS_LEVEL = conf("spark.rapids.sql.metrics.level")
+      .doc("GPU plans can produce a lot more metrics than CPU plans do. In very large " +
+          "queries this can sometimes result in going over the max result size limit for the " +
+          "driver. Supported values include DEBUG: which will enable all metrics supported and " +
```
Nit: Suggest removing the `:` from the values since it's not immediately clear whether it's part of the value.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:
```diff
@@ -410,6 +429,21 @@ object RapidsConf {
+          "queries this can sometimes result in going over the max result size limit for the " +
+          "driver. Supported values include DEBUG: which will enable all metrics supported and " +
+          "typically only needs to be enabled when debugging the plugin. MODERATE: which is the " +
+          "default value and should output enough metrics to understand how long each part of " +
```
Nit: Don't think it's necessary to mention the default value since it appears both here and in the generated documentation next to this text.

revans2 commented on shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuBroadcastHashJoinExec.scala:
```diff
@@ -1,11 +1,11 @@
  * You may obtain a copy of the License at
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *   http://www.apache.org/licenses/LICENSE-2.0
```
I did it to match spark301 and spark301db versions of the same file. I picked wrong on which version to fix, so I will switch it.

revans2 commented:
build

revans2 commented:
@jlowe I think I have addressed all of your review comments. I also updated some of the code that made some metrics optional with nulls and instead turned them into NoopMetric instances.

abellina commented on shims/spark301db/src/main/scala/com/nvidia/spark/rapids/shims/spark301db/GpuShuffledHashJoinExec.scala:
```diff
@@ -85,13 +82,15 @@ case class GpuShuffledHashJoinExec(
+  override lazy val additionalMetrics: Map[String, GpuMetric] = Map(
+    BUILD_DATA_SIZE -> createSizeMetric(ESSENTIAL_LEVEL, DESCRIPTION_BUILD_DATA_SIZE),
+    BUILD_TIME -> createNanoTimingMetric(ESSENTIAL_LEVEL, DESCRIPTION_BUILD_TIME),
+    STREAM_TIME -> createNanoTimingMetric(MODERATE_LEVEL, STREAM_TIME),
```
should this be `DESCRIPTION_STREAM_TIME` on the right?

abellina replied:
same for `DESCRIPTION_JOIN_TIME` below?

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarToGpu.scala:
```diff
@@ -387,14 +384,14 @@ case class HostColumnarToGpu(child: SparkPlan, goal: CoalesceGoal)
+    val numInputBatches = gpuLongMetric(NUM_INPUT_BATCHES)
+    val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
+    val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
+    val collectTime = gpuLongMetric("collectTime")
```
can we use `COLLECT_TIME` and `CONCAT_TIME` here?

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileSourceScanExec.scala:
```diff
@@ -342,36 +343,39 @@ case class GpuFileSourceScanExec(
   override protected def doExecuteColumnar(): RDD[ColumnarBatch] = {
-    val numOutputRows = longMetric("numOutputRows")
-    val scanTime = longMetric("scanTime")
+    val numOutputRows = gpuLongMetric("numOutputRows")
```
"numOutputRows" => `NUM_OUTPUT_ROWS`

revans2 commented:
build

revans2 commented:
Thanks for the review @abellina I think I have addressed all of your concerns
